### PR TITLE
Mo sg tracking list item updates

### PR DESCRIPTION
**Pull request by smguerra**
## Description

This PR adds functionality to the ListItem component that allows users to mark their items as purchased and track what they do and do not need to buy. The feature includes a checkbox with a semantic label that, when checked, updates the dateLastPurchased and totalPurchases properties on the corresponding Firestore document. The item is shown as checked for 24 hours after the purchase is made, after which it unchecks itself so the user can buy it again.

The implementation includes updates to multiple parts of the Firestore document, which are handled by the updateItem function in firebase.js. The feature has been tested using the Firestore console to verify that the dateLastPurchased property is updated correctly.

## Changes Made
- Added checkbox with semantic label to ListItem component
- Implemented updateItem function in firebase.js to update Firestore document properties
- Added logic to show item as checked for 24 hours after purchase
- Verified feature using Firestore console

## Related Issue

Closes issue https://github.com/the-collab-lab/tcl-59-smart-shopping-list/issues/8

## Acceptance Criteria

- [x] The ListItem component renders a checkbox with a semantic <label>.
- [x] Checking off the item in the UI also updates the dateLastPurchased and totalPurchases properties on the corresponding Firestore document.
- [x] The item is shown as checked for 24 hours after the purchase is made (i.e. we assume the user does not need to buy the item again for at least 1 day). After 24 hours, the item unchecks itself so the user can buy it again.
- [x] The updateItem function in firebase.js has been filled out, and sends updates to the firestore database when an item is checked or un-checked.

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Testing Steps / QA Criteria

* In your terminal, pull this branch down with the command- `git pull origin mo-sg-tracking-listItem-updates` and switch to the branch with `git checkout mo-sg-tracking-listItem-updates` 
* You will need to run `npm ci` to install all dependencies or run `npm i date-fns` to install the date-fns dependency we used in the project.
* Then `npm start` to launch the app.
* Navigate to the list page in the app
* Check the box next to the item to mark it as purchased
* Verify that the dateLastPurchased and totalPurchases properties on the Firestore document for the item have been updated
* Wait 24 hours and check the box again
* Verify that the item has un-checked itself and can be purchased again

### CODE REVIEW

**Comment by KarenJParker77** [comment]
Hello Sandra and Mike 😃 
This is very impressive. (I am looking forward to hearing how it works on Monday!) I have put a check in next to an item, and it has updated Firestore accordingly. I will check it again tomorrow to make sure it resets. 

One question I have is once I checked the box, I was able to uncheck it again myself - should the tick stay in place for 24 hours? Firestore shows i've bought the item 20 times, which I haven't, i've just ticked and unticked a few times!

**Comment by smguerra** [comment]
Hello Karen, 

Great! many thanks for your review, we will look into it.  the functionallity is yet to be finished,, this pull request is still a work in progress, we will let you know when its ready for review,

**Comment by KarenJParker77** [comment]
Ohhhhhh, so it is! Excuse my eagerness 🤦‍♀️🤣

**Comment by smguerra** [comment]
Hello Karen,

No worries at all! Your review is very much appreciated. 😊
As for your question, the ticking should stay in place for 24 hours once checked, so we'll make sure to update the code to reflect this. Thank you for bringing this to our attention.

Please, let us know if you have any further questions.

**Comment by mikeyxx** [comment]
> Hello Sandra and Mike 😃 This is very impressive. (I am looking forward to hearing how it works on Monday!) I have put a check in next to an item, and it has updated Firestore accordingly. I will check it again tomorrow to make sure it resets.
> 
> One question I have is once I checked the box, I was able to uncheck it again myself - should the tick stay in place for 24 hours? Firestore shows i've bought the item 20 times, which I haven't, i've just ticked and unticked a few times!

Hey Karen, the feature has now been completed. You can test it out. 
For the issues you raised previously, once you check an item in the list, it updates the firebase store and the item stays checked for 24 hours after which it unchecks. You can manually uncheck an item but if the item has not been bought for more than 24 hours, if the page refreshes again, the item would remain checked. Also we have fixed the bug on the total purchase counter

**Comment by mikeyxx** [comment]
> > Hello Sandra and Mike 😃 This is very impressive. (I am looking forward to hearing how it works on Monday!) I have put a check in next to an item, and it has updated Firestore accordingly. I will check it again tomorrow to make sure it resets.
> > One question I have is once I checked the box, I was able to uncheck it again myself - should the tick stay in place for 24 hours? Firestore shows i've bought the item 20 times, which I haven't, i've just ticked and unticked a few times!
> 
> Hey Karen, the feature has now been completed. You can test it out. For the issues you raised previously, once you check an item in the list, it updates the firebase store and the item stays checked for 24 hours after which it unchecks. You can manually uncheck an item but if the item has not been bought for more than 24 hours, if the page refreshes again, the item would remain checked. Also we have fixed the bug on the total purchase counter

Hey @KarenJParker77, because you asked, we have disabled the checkbox so that once an item is checked, you can no longer manually uncheck it. It unchecks it's self after 24 hours.😊

Thanks for your suggestion

File: src/App.jsx
Comment (edieblu) [inline src/App.jsx:59]: ❤️

File: package.json
Comment (edieblu) [inline package.json:10]: Yes, very good! 👏
It's a good idea to handle everything regarding date and time and, in particular, the difference between two dates or times with a library, it's very easy to make a mistake otherwise (we just had a bug at work where someone did this manually and things broke for us on March 1 as the system was expecting February to have at least 30 days...)

File: src/api/firebase.js
Comment (edieblu) [inline src/api/firebase.js]: A general suggestion for async functions is to wrap them in a try/catch block for better error handling. 😊

File: src/components/ListItem.jsx
Comment (edieblu) [inline src/components/ListItem.jsx:14]:
```suggestion
```

File: src/components/ListItem.jsx
Comment (edieblu) [inline src/components/ListItem.jsx]: 💡 Instead of doing this twice you could store the result of calling this function into a variable, e.g. isRecentlyPurchased, and then you could simplify your props like this: 

```
checked={!isRecentlyPurchased}
disabled={isRecentlyPurchased}
```

File: src/components/ListItem.jsx
Comment (edieblu) [inline src/components/ListItem.jsx:13]: This is an async function, please wrap it in try/catch

File: src/components/ListItem.jsx
Comment (edieblu) [inline src/components/ListItem.jsx]: 💡 For a bonus point, when the checkbox is disabled, add a tooltip explaining why it is disabled (hint, checkout the `title` attribute).

**Review by edieblu — CHANGES_REQUESTED** [review]
Hello team 🕺, 
Good job! It's already looking very good!!
The functionality is there, and I added a couple of pointers to make the code even better!

File: src/components/ListItem.jsx
Comment (mikeyxx) [reply src/components/ListItem.jsx:13]: The logic that calls the updateItem is currently wrapped inside a try/catch block

File: package.json
Comment (mikeyxx) [reply package.json:10]: 🕺😅

File: src/api/firebase.js
Comment (mikeyxx) [reply src/api/firebase.js]: Thanks. try/catch block has been added

File: src/components/ListItem.jsx
Comment (mikeyxx) [reply src/components/ListItem.jsx]: Done.

File: src/components/ListItem.jsx
Comment (mikeyxx) [reply src/components/ListItem.jsx]: Nice. It's adds a good user experience

File: src/components/ListItem.jsx
Comment (edieblu) [reply src/components/ListItem.jsx:13]: You're right, I missed that 👍
I expected the try to wrap the `updateItem` directly, but this is okay too

File: src/components/ListItem.jsx
Comment (edieblu) [inline src/components/ListItem.jsx]: image

File: src/components/ListItem.jsx
Comment (edieblu) [inline src/components/ListItem.jsx]: I think you should only add the `title` attribute once, and since the `input` is the child of this `li` (and it also already has a `label`), let's just add it the `li` and remove it for the `input`. 😊

**Comment by edieblu** [comment]
Hiya @mikeyxx and @smguerra 
really good work on getting through my comments so quickly ⚡ 
I added a couple more comments - we're good to go once we resolve this! 🚀
Oh, one last thing, can you make sure that any TO DO comments that have already been resolved are deleted from the code 😊

Before merging to `main` you'll also have to resolve a git conflict - let me know if you need any help with that!

File: src/components/ListItem.jsx
Comment (mikeyxx) [reply src/components/ListItem.jsx]: Ah missed that. It's resolved

File: src/components/ListItem.jsx
Comment (mikeyxx) [reply src/components/ListItem.jsx]: I thought I hovered on the label and a description didn't show when I had it on the label so I just added the title on the two. I've updated it now. 😅

**Review by edieblu — APPROVED** [review]
Woot-woot! 🚆

**Comment by mikeyxx** [comment]
@edieblu Thank you for the comments and recommendations. Learnt so much. On your last comment about removing the Todo comments, I have intentionally left them in for documentation purposes so that I can go back to the code after we are done to understand all the tasks and steps we took to build the complete application.

**Pull request merged**
